doc/man/man3/lis_matvec.3
.TH lis_matvec 3 "6 Sep 2012" "Man Page" "Lis Library Functions"

.SH NAME

lis_matvec \- calculate the matrix vector product

.SH SYNOPSIS

#include "lis.h"

\fBLIS_INT lis_matvec\fR(\fBLIS_MATRIX A\fR, \fBLIS_VECTOR x\fR, \fBLIS_VECTOR y\fR);

.SH DESCRIPTION

Calculate the matrix vector product \fIy = Ax\fR.

.SH INPUT

.IP "\fBA\fR"
The matrix

.IP "\fBx\fR"
The vector

.SH OUTPUT

.IP "\fBy\fR"
\fIAx\fR

.SH EXIT STATUS

The following exit values are returned:
.IP "0"
The process is normally terminated
.IP "unspecified"
An error occurred

.SH SEE ALSO

.BR lis (3)
.PP
http://www.ssisc.org/lis/



ChangeLog
Version 1.5.40  December 20, 2014

  - Fixed variable declaration bug in lis_precision_vec.c.

Version 1.5.39  December 16, 2014

  - Updated documents.

Version 1.5.38  December 15, 2014

  - Updated documents.

Version 1.5.37  December 14, 2014

  - Cleaned up programs (thanks to Florian Lehner).

Version 1.5.36  December 11, 2014

  - Updated lis_esolver_rqi.c.

Version 1.5.35  December 10, 2014

  - Updated documents (thanks to Sushilkumar Sreekumar).

Version 1.5.34  December 8, 2014

  - Updated documents.
  - Changed default preconditioners for eigensolvers.

Version 1.5.33  December 7, 2014

  - Updated documents.
  - Added experimental support for Jacobi-Davidson.
  - Fixed exception handling bug in lis_matrix_set_size() (thanks to Yi Yao).

Version 1.5.32  December 5, 2014

  - Updated documents.

Version 1.5.31  December 4, 2014

  - Updated etest7.c.
  - Updated lis_esolver_ai.c.

Version 1.5.30  December 3, 2014

  - Updated documents.
  - Fixed print bug on iteration counts of inner eigensolvers.

Version 1.5.29  December 2, 2014

  - Added experimental support for Arnoldi iteration.
  - Disabled obsolete option -m for eigensolvers.
  - Updated documents.

Version 1.5.28  December 1, 2014

  - Changed default linear solver for eigensolvers.
  - Changed order of eigensolvers:

     flags      | eigensolvers              flags     | eigensolvers
    ------------+--------------------     ------------+--------------------
     -e {si|5}  | Subspace            ==>  -e {cg|5}  | CG
     -e {li|6}  | Lanczos                  -e {cr|6}  | CR
     -e {cg|7}  | CG                       -e {si|7}  | Subspace
     -e {cr|8}  | CR                       -e {li|8}  | Lanczos

  - Updated documents.

Version 1.5.27  November 30, 2014

  - Updated documents.

Version 1.5.26  November 29, 2014

  - Changed names of functions lis_array_matvec2() and
    lis_array_matmat2() to lis_array_matvec_ns() and lis_array_matmat_ns(),
    respectively.

Version 1.5.25  November 28, 2014

  - Updated lis_array.c.

Version 1.5.24  November 27, 2014

  - Updated lis_array.c (thanks to Florian Lehner).

Version 1.5.23  November 26, 2014

  - Cleaned up lis_esolver_li.c.
  - Cleaned up lis_array.c.
  - Fixed convergence condition of lis_array_qr().

Version 1.5.22  November 25, 2014

  - Updated documents.

Version 1.5.21  November 24, 2014

  - Cleaned up lis_solver_*.c.
  - Added support for test6f.F90 for array operations.
  - Updated documents.

Version 1.5.20  November 23, 2014

  - Updated documents.

Version 1.5.19  November 22, 2014

  - Fixed bug in lisf_array.c (thanks to Florian Lehner).

Version 1.5.18  November 21, 2014

  - Updated documents.

Version 1.5.17  November 20, 2014

  - Updated lis_fortran.h for Win64 environment.
  - Fixed initial value related bug in GMRES to enable option -initx_zeros.

Version 1.5.16  November 19, 2014

  - Updated documents (thanks to Sushilkumar Sreekumar).

Version 1.5.15  November 18, 2014

  - Fixed order of array elements in lis_array_cgs(), lis_array_mgs(),
    and lis_array_qr().
  - Fixed memory initialization bug in lis_array_mgs().
  - Updated documents.

Version 1.5.14  November 16, 2014

  - Cleaned up programs.
  - Updated eigensolvers.
  - Added sample scripts for graphics.

Version 1.5.13  November 14, 2014

  - Updated documents.

Version 1.5.12  November 13, 2014

  - Updated documents.

Version 1.5.11  November 12, 2014

  - Updated test3b.c.

Version 1.5.10  November 11, 2014

  - Added sample scripts for graphics.

Version 1.5.9  November 9, 2014

  - Cleaned up programs.

Version 1.5.8  November 8, 2014

  - Updated documents.

Version 1.5.7  November 7, 2014

  - Fixed print format.
  - Fixed elapsed time related bug in preconditioned eigensolvers.
  - Fixed CR related bug in lis_esolver_cg.c.

Version 1.5.6  November 6, 2014

  - Updated documents.

Version 1.5.5  November 5, 2014

  - Cleaned up programs.
  - Updated documents.

Version 1.5.4  November 4, 2014

  - Fixed quad precision related bug in lis_esolver_si.c.
  - Fixed bug in test/Makefile.am.

Version 1.5.3  November 3, 2014

  - Updated eigensolvers.
  - Fixed array related bug in lis_esolver_cg.c.

Version 1.5.2  November 1, 2014

  - Added support for etest7.c for array operations.
  - Changed specification of lis_array_qr() to return number of iterations
    and error of QR algorithm.
  - Updated documents.

Version 1.5.1  October 31, 2014

  - Added support for test6.c for array operations.
  - Updated print formats.
  - Updated documents.

Version 1.5.0  October 30, 2014

  - Fixed print bug in option -conv_cond (thanks to Suguru Miyauchi).
  - Added official support for array operations.
  - Updated documents.

Version 1.4.67  October 28, 2014

  - Fixed bug in lis_array_nrmi_f() (thanks to Florian Lehner).
  - Updated lis_array.c.

Version 1.4.66  October 27, 2014

  - Updated lis_array.c.
  - Changed names of functions lis_array_dot2() and
    lis_array_invGauss() to lis_array_dot() and lis_array_ge(),
    respectively.

Version 1.4.65  October 23, 2014

  - Updated documents (thanks to Suguru Miyauchi).

Version 1.4.64  October 21, 2014

  - Updated lis_m_data_creation_AMGCG.F90.
  - Updated sample scripts for graphics.

Version 1.4.63  October 19, 2014

  - Added sample scripts for graphics (thanks to Takashi Shiroto
    and Sushilkumar Sreekumar).

Version 1.4.62  October 17, 2014

  - Fixed exception handling bug in lis_m_data_creation_AMGCG.F90
    (thanks to Takashi Shiroto).

Version 1.4.61  September 19, 2014

  - Fixed OpenMP related bug in lis_matrix_ell.c (thanks to Gerald Wedel).

Version 1.4.60  September 13, 2014

  - Fixed long double related print bugs.
  - Fixed long long int related print bugs.

Version 1.4.59  September 9, 2014

  - Updated configure.in.
  - Updated test programs in Fortran.
  - Updated documents.

Version 1.4.58  August 27, 2014

  - Added support for lis_matrix_unset() in Fortran (thanks to Takumi Yanagawa

src/matvec/lis_matvec.c
/* Copyright (C) 2005 The Scalable Software Infrastructure Project. All rights reserved.

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions are met:
   1. Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
   2. Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
   3. Neither the name of the project nor the names of its contributors 
      may be used to endorse or promote products derived from this software 
      without specific prior written permission.

   THIS SOFTWARE IS PROVIDED BY THE SCALABLE SOFTWARE INFRASTRUCTURE PROJECT
   ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
   TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
   PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE SCALABLE SOFTWARE INFRASTRUCTURE
   PROJECT BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
   OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
   SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
   INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
   ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
   POSSIBILITY OF SUCH DAMAGE.
*/

#ifdef HAVE_CONFIG_H
	#include "lis_config.h"
#else
#ifdef HAVE_CONFIG_WIN_H
	#include "lis_config_win.h"
#endif
#endif

#include <stdio.h>
#include <stdlib.h>
#ifdef HAVE_MALLOC_H
        #include <malloc.h>
#endif
#include <string.h>
#include <stdarg.h>
#ifdef _OPENMP
	#include <omp.h>
#endif
#ifdef USE_MPI
	#include <mpi.h>
#endif
#include "lislib.h"

LIS_MATVEC_FUNC LIS_MATVEC  = lis_matvec;
LIS_MATVEC_FUNC LIS_MATVECT = lis_matvect;

#undef __FUNC__
#define __FUNC__ "lis_matvec"
LIS_INT lis_matvec(LIS_MATRIX A, LIS_VECTOR X, LIS_VECTOR Y)
{
	LIS_SCALAR *x,*y;

	LIS_DEBUG_FUNC_IN;

	if( X->precision==LIS_PRECISION_DEFAULT )
	{
		x = X->value;
		y = Y->value;

		switch( A->matrix_type )
		{
		case LIS_MATRIX_CSR:
			#ifdef USE_MPI
				LIS_MATVEC_SENDRECV;
			#endif
			lis_matvec_csr(A, x, y);
			break;
		case LIS_MATRIX_BSR:
			#ifdef USE_MPI
				LIS_MATVEC_SENDRECV;
			#endif
			if( A->bnr<=4 && A->bnc<=4 )
			{
				lis_matvec_bsr_xxx[A->bnr-1][A->bnc-1](A, x, y);
			}
			else
			{
				lis_matvec_bsr(A, x, y);
			}
			break;
		case LIS_MATRIX_CSC:
			#ifdef USE_MPI
				LIS_MATVEC_SENDRECV;
			#endif
			lis_matvec_csc(A, x, y);
			break;
		case LIS_MATRIX_BSC:
			#ifdef USE_MPI
				LIS_MATVEC_SENDRECV;
			#endif
			lis_matvec_bsc(A, x, y);
			break;
		case LIS_MATRIX_MSR:
			#ifdef USE_MPI
				LIS_MATVEC_SENDRECV;
			#endif
			lis_matvec_msr(A, x, y);
			break;
		case LIS_MATRIX_ELL:
			#ifdef USE_MPI
				LIS_MATVEC_SENDRECV;
			#endif
			lis_matvec_ell(A, x, y);
			break;
		case LIS_MATRIX_DIA:
			#ifdef USE_MPI
				LIS_MATVEC_SENDRECV;
			#endif
			lis_matvec_dia(A, x, y);
			break;
		case LIS_MATRIX_JAD:
			#ifdef USE_MPI
				#ifndef USE_OVERLAP
					LIS_MATVEC_SENDRECV;
				#else
					LIS_MATVEC_REALLOC;
				#endif
			#endif
			lis_matvec_jad(A, x, y);
			break;
		case LIS_MATRIX_VBR:
			#ifdef USE_MPI
				LIS_MATVEC_SENDRECV;
			#endif
			lis_matvec_vbr(A, x, y);
			break;
		case LIS_MATRIX_DNS:
			#ifdef USE_MPI
				LIS_MATVEC_SENDRECV;
			#endif
			lis_matvec_dns(A, x, y);
			break;
		case LIS_MATRIX_COO:
			#ifdef USE_MPI
				LIS_MATVEC_SENDRECV;
			#endif
			lis_matvec_coo(A, x, y);
			break;
		default:
			LIS_SETERR_IMP;
			return LIS_ERR_NOT_IMPLEMENTED;
			break;
		}
	}
#ifdef USE_QUAD_PRECISION
	else
	{

		switch( A->matrix_type )
		{
		case LIS_MATRIX_CSR:
			#ifdef USE_MPI
				lis_send_recv_mp(A->commtable,X);
			#endif
			#ifndef USE_FMA2_SSE2
				lis_matvec_csr_mp(A, X, Y);
			#else
				lis_matvec_csr_mp2(A, X, Y);
			#endif
			break;
		case LIS_MATRIX_CSC:
			#ifdef USE_MPI
				lis_send_recv_mp(A->commtable,X);
			#endif
			#ifndef USE_FMA2_SSE2
				lis_matvec_csc_mp(A, X, Y);
			#else
				lis_matvec_csc_mp2(A, X, Y);
			#endif
			break;
		default:
			LIS_SETERR_IMP;
			return LIS_ERR_NOT_IMPLEMENTED;
			break;
		}
	}
#endif

	LIS_DEBUG_FUNC_OUT;
	return LIS_SUCCESS;
}

#undef __FUNC__
#define __FUNC__ "lis_matvect"
LIS_INT lis_matvect(LIS_MATRIX A, LIS_VECTOR X, LIS_VECTOR Y)
{
	LIS_SCALAR *x,*y;

	LIS_DEBUG_FUNC_IN;

	x = X->value;
	y = Y->value;

	if( X->precision==LIS_PRECISION_DEFAULT )
	{
		switch( A->matrix_type )
		{
		case LIS_MATRIX_CSR:
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE0;
			#endif
			lis_matvect_csr(A, x, y);
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE;
			#endif
			break;
		case LIS_MATRIX_BSR:
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE0;
			#endif
			lis_matvect_bsr(A, x, y);
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE;
			#endif
			break;
		case LIS_MATRIX_CSC:
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE0;
			#endif
			lis_matvect_csc(A, x, y);
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE;
			#endif
			break;
		case LIS_MATRIX_BSC:
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE0;
			#endif
			lis_matvect_bsc(A, x, y);
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE;
			#endif
			break;
		case LIS_MATRIX_MSR:
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE0;
			#endif
			lis_matvect_msr(A, x, y);
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE;
			#endif
			break;
		case LIS_MATRIX_ELL:
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE0;
			#endif
			lis_matvect_ell(A, x, y);
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE;
			#endif
			break;
		case LIS_MATRIX_JAD:
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE0;
			#endif
			lis_matvect_jad(A, x, y);
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE;
			#endif
			break;
		case LIS_MATRIX_DIA:
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE0;
			#endif
			lis_matvect_dia(A, x, y);
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE;
			#endif
			break;
		case LIS_MATRIX_VBR:
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE0;
			#endif
			lis_matvect_vbr(A, x, y);
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE;
			#endif
			break;
		case LIS_MATRIX_DNS:
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE0;
			#endif
			lis_matvect_dns(A, x, y);
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE;
			#endif
			break;
		case LIS_MATRIX_COO:
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE0;
			#endif
			lis_matvect_coo(A, x, y);
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE;
			#endif
			break;
		default:
			LIS_SETERR_IMP;
			return LIS_ERR_NOT_IMPLEMENTED;
			break;
		}
	}
#ifdef USE_QUAD_PRECISION
	else
	{
		switch( A->matrix_type )
		{
		case LIS_MATRIX_CSR:
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE0;
			#endif
			#ifndef USE_FMA2_SSE2
				lis_matvect_csr_mp(A, X, Y);
			#else
				lis_matvect_csr_mp2(A, X, Y);
			#endif
			#ifdef USE_MPI
				lis_reduce_mp(A->commtable,Y);
			#endif
			break;
		case LIS_MATRIX_CSC:
			#ifdef USE_MPI
				LIS_MATVEC_REDUCE0;
			#endif
			#ifndef USE_FMA2_SSE2
				lis_matvect_csc_mp(A, X, Y);
			#else
				lis_matvect_csc_mp2(A, X, Y);
			#endif
			#ifdef USE_MPI
				lis_reduce_mp(A->commtable,Y);
			#endif
			break;
		default:
			LIS_SETERR_IMP;
			return LIS_ERR_NOT_IMPLEMENTED;
			break;
		}
	}
#endif
	LIS_DEBUG_FUNC_OUT;
	return LIS_SUCCESS;
}



#undef __FUNC__
#define __FUNC__ "lis_matvec_optimize"
LIS_INT lis_matvec_optimize(LIS_MATRIX A, LIS_INT *matrix_type_maxperf)
{
  LIS_INT	   nprocs,my_rank;
  int    	   int_nprocs,int_my_rank;
  LIS_INT	   err,iter,i,matrix_type,ss,se;
  double	   time,time2,convtime;
  LIS_SCALAR       val;
  double	   commtime,comptime,flops,flops_maxperf;
  LIS_MATRIX       A1;
  LIS_VECTOR       X, Y;
  char             *lis_storagename2[]   = {"CSR", "CSC", "MSR", "DIA", "ELL", "JAD", "BSR", "BSC", "VBR", "COO", "DNS"};

	LIS_DEBUG_FUNC_IN;

#ifdef USE_MPI
	MPI_Comm_size(MPI_COMM_WORLD,&int_nprocs);
	MPI_Comm_rank(MPI_COMM_WORLD,&int_my_rank);
	nprocs = int_nprocs;
	my_rank = int_my_rank;
#else
	nprocs  = 1;
	my_rank = 0;
#endif

	ss = 1;
	se = 11;

	lis_vector_duplicate(A,&X);
	lis_vector_duplicate(A,&Y);
	err = lis_vector_set_all(1.0,X);

	printf("\nmeasuring matvec performance...\n");
	iter = (int)(10000000 / A->nnz) + 1; 
	flops_maxperf = 0.0;
#ifdef _LONGLONG
	printf("number of iterations = 1e7 / %lld + 1 = %lld\n", A->nnz, iter);
#else
	printf("number of iterations = 1e7 / %d + 1 = %d\n", A->nnz, iter);
#endif
	for (matrix_type=ss;matrix_type<se;matrix_type++)
	  {
	    if ( nprocs>1 && matrix_type==9 ) continue;
	    lis_matrix_duplicate(A,&A1);
	    lis_matrix_set_type(A1,matrix_type);
	    err = lis_matrix_convert(A,A1);
	    if( err ) CHKERR(err);
		    
	    comptime = 0.0;
	    commtime = 0.0;

	    for(i=0;i<iter;i++)
	      {
#ifdef USE_MPI
		MPI_Barrier(A1->comm);
		time = lis_wtime();
		lis_send_recv(A1->commtable,X->value);
		commtime += lis_wtime() - time;
#endif
		time2 = lis_wtime();
		lis_matvec(A1,X,Y);
		comptime += lis_wtime() - time2;
	      }
	    lis_vector_nrm2(Y,&val);

	    if( my_rank==0 )
	      {
		flops = 2.0*A->nnz*iter*1.0e-6 / comptime;
#ifdef USE_MPI
#ifdef _LONG__DOUBLE
#ifdef _LONGLONG
		printf("matrix_type = %2lld (%s), computation = %e sec, %8.3f MFLOPS, communication = %e sec, communication/computation = %3.3f\n",matrix_type,lis_storagename2[matrix_type-1],comptime,flops,commtime,commtime/comptime*100);
#else
		printf("matrix_type = %2d (%s), computation = %e sec, %8.3f MFLOPS, communication = %e sec, communication/computation = %3.3f\n",matrix_type,lis_storagename2[matrix_type-1],comptime,flops,commtime,commtime/comptime*100);
#endif
#else
#ifdef _LONGLONG
		printf("matrix_type = %2lld (%s), computation = %e sec, %8.3f MFLOPS, communication = %e sec, communication/computation = %3.3f\n",matrix_type,lis_storagename2[matrix_type-1],comptime,flops,commtime,commtime/comptime*100);
#else
		printf("matrix_type = %2d (%s), computation = %e sec, %8.3f MFLOPS, communication = %e sec, communication/computation = %3.3f\n",matrix_type,lis_storagename2[matrix_type-1],comptime,flops,commtime,commtime/comptime*100);
#endif
#endif
#else
#ifdef _LONG__DOUBLE
#ifdef _LONGLONG
		printf("matrix_type = %2lld (%s), computation = %e sec, %8.3f MFLOPS\n",matrix_type,lis_storagename2[matrix_type-1],comptime,flops);
#else
		printf("matrix_type = %2d (%s), computation = %e sec, %8.3f MFLOPS\n",matrix_type,lis_storagename2[matrix_type-1],comptime,flops);
#endif
#else
#ifdef _LONGLONG
		printf("matrix_type = %2lld (%s), computation = %e sec, %8.3f MFLOPS\n",matrix_type,lis_storagename2[matrix_type-1],comptime,flops);
#else
		printf("matrix_type = %2d (%s), computation = %e sec, %8.3f MFLOPS\n",matrix_type,lis_storagename2[matrix_type-1],comptime,flops);
#endif
#endif
#endif
	      }
	    lis_matrix_destroy(A1);

	    if ( flops > flops_maxperf )
	      {
		*matrix_type_maxperf=matrix_type;
		flops_maxperf = flops;
	      }
	  }

	printf("matrix format is set to %s\n\n", lis_storagename2[*matrix_type_maxperf-1]);

	LIS_DEBUG_FUNC_OUT;
	return LIS_SUCCESS;
}
